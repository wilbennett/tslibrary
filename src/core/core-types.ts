export type NumberArray
    = number[]
    | Float32Array
    | Float64Array
    // | BigInt64Array
    // | BigUint64Array
    | Int8Array
    | Int16Array
    | Int32Array
    | Uint8Array
    | Uint16Array
    | Uint32Array
    | Uint8ClampedArray;

export interface BoundsLike {
    x: number;
    y: number;
    w: number;
    h: number;
}

export function isNumberArray(obj: any): obj is NumberArray {
    return Array.isArray(obj)
        || obj instanceof Float32Array
        || obj instanceof Float64Array
        || obj instanceof Int8Array
        || obj instanceof Int16Array
        || obj instanceof Int32Array
        || obj instanceof Uint8Array
        || obj instanceof Uint16Array
        || obj instanceof Uint32Array
        || obj instanceof Uint8ClampedArray;
}