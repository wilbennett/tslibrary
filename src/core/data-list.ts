import { isTypedArray, NumberArray } from '.';

export class DataList {
  constructor(count: number, elementCount: number);
  constructor(data: NumberArray, elementCount: number);
  constructor(param1: number | NumberArray, elementCount: number) {
    if (typeof param1 === "number") {
      this.count = param1;
      this.elementCount = elementCount;
      this.values = new Array<number>(this.count * this.elementCount);
    } else {
      this.values = param1;
      this.elementCount = elementCount;
      this.count = Math.trunc(this.values.length / this.elementCount);
    }
  }

  static get [Symbol.species]() { return this; }

  readonly count: number;
  readonly elementCount: number;
  readonly values: NumberArray;

  clone() {
    const values = isTypedArray(this.values)
      // @ts-ignore - constructor call.
      ? new this.values.constructor(this.values)
      : this.values.slice();

    // @ts-ignore - species pattern.
    const Species = this.constructor[Symbol.species];
    return new Species(values, this.elementCount);
  }

  copyFrom(list: DataList) {
    const itemCount = Math.min(this.values.length, list.values.length);
    const values = this.values;
    const listValues = list.values;

    for (let i = 0; i < itemCount; i++) {
      values[i] = listValues[i];
    }

    return this;
  }
}
