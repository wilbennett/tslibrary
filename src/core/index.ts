export * from "./core-types";
export * from "./mathex";
export * from "./promiseex";
