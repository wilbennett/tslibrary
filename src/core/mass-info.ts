export class MassInfo {
    constructor(mass: number, inertia: number) {
        this.mass = mass;
        this.inertia = inertia;
        this.massInverse = mass !== 0 ? 1 / mass : 0;
        this.inertiaInverse = inertia !== 0 ? 1 / inertia : 0;
    }

    readonly mass: number;
    readonly massInverse: number;
    readonly inertia: number;
    readonly inertiaInverse: number;

    static readonly empty = new MassInfo(0, 0);
}
