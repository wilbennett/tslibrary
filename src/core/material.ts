export type Material = {
  readonly name: string;
  readonly restitution: number;
  readonly density: number;
  readonly staticFriction: number;
  readonly kineticFriction: number;
};

export const DEFAULT_MATERIAL: Material = Object.freeze({
  name: "default",
  restitution: 0.2,
  density: 0,
  staticFriction: 0.5,
  kineticFriction: 0.3
});
