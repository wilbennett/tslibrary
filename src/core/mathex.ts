import { isNumberArray, NumberArray } from '.';

export class MathEx {
  static readonly TWO_PI = 2 * Math.PI;
  static readonly HALF_PI = Math.PI / 2;
  static readonly QUARTER_PI: number = Math.PI / 4;

  static readonly ONE_RADIAN = 180 / Math.PI;
  static readonly ONE_DEGREE = Math.PI / 180;

  static readonly SIN1 = Math.sin(MathEx.ONE_DEGREE);
  static readonly COS1 = Math.cos(MathEx.ONE_DEGREE);
  static readonly SINN1 = Math.sin(-MathEx.ONE_DEGREE);
  static readonly COSN1 = Math.cos(-MathEx.ONE_DEGREE);

  static readonly EPSILON = 0.00001;

  static toRadians(degrees: number) { return degrees * this.ONE_DEGREE; }
  static toDegrees(radians: number) {
    return radians >= 0 ? radians * this.ONE_RADIAN : radians * this.ONE_RADIAN + 360;
  }

  static toFloat(value: string) { return isNaN(+value) ? Number.NaN : parseFloat(value); }
  static toInt(value: string) { return isNaN(+value) ? Number.NaN : parseInt(value); }
  static toIntRound(value: string) { return isNaN(+value) ? Number.NaN : Math.round(this.toFloat(value)); }

  static toHex(value: number) {
    // >>> 0 forces the value to be positive.
    const result = (value >>> 0).toString(16);
    return result.length % 2 ? "0" + result : result;
  }

  static parseHex(value: string) {
    return parseInt(value, 16);
  }

  static wrapRadians(radians: number) {
    // return MathEx.wrap(radians, -Math.PI, Math.PI);

    if (radians >= -Math.PI && radians <= Math.PI)
      return radians;

    return radians - this.TWO_PI * Math.floor((radians + Math.PI) / this.TWO_PI);
  }

  static wrapDegrees(degrees: number) { return MathEx.wrap(degrees, 0, 360); }

  static lerp(start: number, end: number, t: number) { return (end - start) * t + start; }
  static lerpc(start: number, change: number, t: number) { return change * t + start; }

  static clamp(value: number, min: number, max: number) {
    if (value < min) return min;
    if (value > max) return max;
    return value;
  }

  static clampByte(value: number) { return this.clamp(value, 0, 255); }

  static wrapInt(value: number, low: number, high: number) {
    if (value >= low && value <= high)
      return value;

    value = (value - low) % (high - low + 1);
    return value < 0 ? high + value + 1 : low + value;
  }

  static wrap(value: number, low: number, high: number) {
    if (value >= low && value <= high)
      return value;

    const range = high - low;
    return value = (((value - low) % range) + range) % range + low;
  }

  static mapRange(value: number, min: number, max: number, newMin: number, newMax: number) {
    let run = max - min;

    if (run === 0) run = 1;

    const slope = (newMax - newMin) / run;
    const intercept = newMax - (slope * max);
    return value * slope + intercept;
  }

  static sign(value: number) {
    return <any>(value > 0) - <any>(value < 0) || +value;
  }

  static equals(value1: number, value2: number, epsilon: number = this.EPSILON) {
    return Math.abs(value1 - value2) <= epsilon;
  }

  static toFixed(count: number, ...values: number[]) {
    return values.map((v) => v.toFixed(count));
  }

  static random<T>(array: T[]): T;
  static random(): number;
  static random(max: number): number;
  static random(min: number, max: number): number;
  static random<T>(min?: number | T[], max?: number) {
    let array: T[] | null = null;

    if (Array.isArray(min)) {
      array = min;
      min = 0;
      max = array.length - 1;
    } else if (min === undefined) {
      min = 0;
      max = 1;
    } else if (!max) {
      max = min;
      min = 0;
    }

    return array
      ? array[Math.floor(Math.random() * (max - min + 1)) + min]
      : Math.random() * (max - min) + min;
  }

  static randomInt(max: number): number;
  static randomInt(min: number, max: number): number;
  static randomInt(min: number, max?: number) {
    if (max === undefined) {
      max = Math.floor(min);
      min = 0;
    } else {
      min = Math.ceil(min);
      max = Math.floor(max);
    }

    return Math.floor(Math.random() * (max - min + 1)) + min;
  }

  static stddev(values: NumberArray): number;
  static stddev(...values: number[]): number;
  static stddev(param1: any): number {
    const values = isNumberArray(param1) ? param1 : <number[]>Array.from(arguments);

    let sum = 0;
    const length = values.length;

    for (let i = 0; i < length; i++) {
      sum += values[i];
    }

    const mean = sum / length;
    let sqrDiff = 0;

    for (let i = 0; i < length; i++) {
      const diff = values[i] - mean;
      sqrDiff += diff * diff;
    }

    const variance = sqrDiff / length;
    return Math.sqrt(variance);
  }

  static stddevsamp(values: NumberArray): number;
  static stddevsamp(...values: number[]): number;
  static stddevsamp(param1: any): number {
    const values = isNumberArray(param1) ? param1 : <number[]>Array.from(arguments);

    let sum = 0;
    const length = values.length;

    for (let i = 0; i < length; i++) {
      sum += values[i];
    }

    const mean = sum / length;
    let sqrDiff = 0;

    for (let i = 0; i < length; i++) {
      const diff = values[i] - mean;
      sqrDiff += diff * diff;
    }

    const variance = sqrDiff / (length - 1);
    return Math.sqrt(variance);
  }

  static calcGrowth(startValue: number, growthRate: number, time: number) {
    return startValue * Math.pow(1 + growthRate, time);
  }

  static calcDecay(startValue: number, decayRate: number, time: number) {
    return startValue * Math.pow(1 - decayRate, time);
  }

  static calcGrowthTime(startValue: number, decayRate: number, targetValue: number) {
    return Math.log(startValue / targetValue) / Math.log(1 + decayRate);
  }

  static calcDecayTime(startValue: number, decayRate: number, targetValue: number) {
    return Math.log(startValue / targetValue) / Math.log(1 - decayRate);
  }
}
