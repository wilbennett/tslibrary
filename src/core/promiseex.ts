export class PromiseEx<T> {
    private _resolve!: (value?: T | PromiseLike<T>) => void;
    private _reject!: (reason?: any) => void;

    constructor() {
        const self = this;

        this._promise = new Promise<T>((resolve, reject) => {
            self._resolve = resolve;
            self._reject = reject;
        });
    }

    private _promise: Promise<T>;
    get promise() { return this._promise; }

    resolve(value?: T | PromiseLike<T>) { this._resolve(value); }
    reject(reason?: any) { this._reject(reason); }

    then<R>(onfulfilled?: (v: T) => R | PromiseLike<R>, onrejected?: (reason?: any) => PromiseLike<never>) {
        return this._promise.then(onfulfilled, onrejected);
    }

    catch(onrejected?: (reason?: any) => PromiseLike<never>) { return this._promise.catch(onrejected); }
    finally(f: () => void) { return this._promise.finally(f); }
}
