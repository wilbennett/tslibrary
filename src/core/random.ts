const TWO_POW31 = Math.pow(2, 31);
const TWO_POW31_MINUS1 = TWO_POW31 - 1;
const TWO_POW31_MINUS2 = TWO_POW31 - 2;
const TWO_POW31_MINUS2_INV = 1 / TWO_POW31_MINUS2;

// https://gist.github.com/blixt/f17b47c62508be59987b
// http://www.firstpr.com.au/dsp/rand31/
export class Random {
    constructor(seed?: number) {
        this.randomize(seed);
    }

    private _seed: number = 0;
    get seed() { return this._seed; }
    set seed(value) {
        this._seed = value % TWO_POW31_MINUS1;

        if (this._seed <= 0)
            this._seed += TWO_POW31_MINUS2;

        this.nextInt(); // Skip first. First float values are usually low.
    }

    randomize(seed?: number) {
        this.seed = seed !== undefined ? seed : Math.round(Math.random() * TWO_POW31) + 1;
    }

    // [1, 2^31 - 2]
    nextInt() { return this._seed = this._seed * 16807 % TWO_POW31_MINUS1; }
    // [0, 1)
    next() { return (this.nextInt() - 1) * TWO_POW31_MINUS2_INV; }
}
