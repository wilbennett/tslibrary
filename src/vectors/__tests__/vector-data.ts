import { VectorDataX, VectorDataXYW, VectorDataXYZW } from '..';
import { DataList } from '../../core';

describe("Should allow working with vector data", () => {
  const count = 2;

  function createX() { return new VectorDataX(count); }
  function createXYW() { return new VectorDataXYW(count); }
  function createXYZW() { return new VectorDataXYZW(count); }
  function createArrayX() { return new VectorDataX(new Array<number>(count)); }
  function createArrayXYW() { return new VectorDataXYW(new Array<number>(count * 3)); }
  function createArrayXYZW() { return new VectorDataXYZW(new Array<number>(count * 4)); }

  function createFloat32X() { return new VectorDataX(new Float32Array(count)); }
  function createFloat32XYW() { return new VectorDataXYW(new Float32Array(count * 3)); }
  function createFloat32XYZW() { return new VectorDataXYZW(new Float32Array(count * 4)); }

  describe("Should allow creating instances", () => {
    it.each([
      ["vector data X with count", createX()],
      ["vector data X with existing", createArrayX()],
      ["vector data XYW with count", createXYW()],
      ["vector data XYW with existing", createArrayXYW()],
      ["vector data XYZW with count", createXYZW()],
      ["vector data XYZW with existing", createArrayXYZW()],
      ["vector data X with typed array", createFloat32X()],
      ["vector data XYW with typed array", createFloat32XYW()],
      ["vector data XYZW with typed array", createFloat32XYZW()]
    ])(
      "Should allow creating %s",
      (_, l) => {
        const list = <DataList>l;

        expect(list.count).toBe(count);
        expect(list.elementCount).toMatchSnapshot();
        expect(list.values.length).toMatchSnapshot();
      });
  });

  describe("Should calculate start index", () => {
    it.each([
      ["vector data X with count", createX()],
      ["vector data X with existing", createArrayX()],
      ["vector data XYW with count", createXYW()],
      ["vector data XYW with existing", createArrayXYW()],
      ["vector data XYZW with count", createXYZW()],
      ["vector data XYZW with existing", createArrayXYZW()],
      ["vector data X with typed array", createFloat32X()],
      ["vector data XYW with typed array", createFloat32XYW()],
      ["vector data XYZW with typed array", createFloat32XYZW()]
    ])(
      "Should calculate for %s",
      (_, l) => {
        const list = <DataList>l;

        expect(list.getStartIndex(0)).toBe(0);
        expect(list.getStartIndex(1)).toMatchSnapshot();
        expect(list.getStartIndex(2)).toMatchSnapshot();
        expect(list.getStartIndex(3)).toMatchSnapshot();
      });
  });
});
