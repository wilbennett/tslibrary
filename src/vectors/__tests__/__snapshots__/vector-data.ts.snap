// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data X with count 1`] = `1`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data X with count 2`] = `2`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data X with existing 1`] = `1`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data X with existing 2`] = `2`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data X with typed array 1`] = `1`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data X with typed array 2`] = `2`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data XYW with count 1`] = `3`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data XYW with count 2`] = `6`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data XYW with existing 1`] = `3`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data XYW with existing 2`] = `6`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data XYW with typed array 1`] = `3`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data XYW with typed array 2`] = `6`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data XYZW with count 1`] = `4`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data XYZW with count 2`] = `8`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data XYZW with existing 1`] = `4`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data XYZW with existing 2`] = `8`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data XYZW with typed array 1`] = `4`;

exports[`Should allow working with vector data Should allow creating instances Should allow creating vector data XYZW with typed array 2`] = `8`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data X with count 1`] = `1`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data X with count 2`] = `2`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data X with count 3`] = `3`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data X with existing 1`] = `1`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data X with existing 2`] = `2`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data X with existing 3`] = `3`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data X with typed array 1`] = `1`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data X with typed array 2`] = `2`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data X with typed array 3`] = `3`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYW with count 1`] = `3`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYW with count 2`] = `6`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYW with count 3`] = `9`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYW with existing 1`] = `3`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYW with existing 2`] = `6`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYW with existing 3`] = `9`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYW with typed array 1`] = `3`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYW with typed array 2`] = `6`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYW with typed array 3`] = `9`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYZW with count 1`] = `4`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYZW with count 2`] = `8`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYZW with count 3`] = `12`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYZW with existing 1`] = `4`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYZW with existing 2`] = `8`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYZW with existing 3`] = `12`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYZW with typed array 1`] = `4`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYZW with typed array 2`] = `8`;

exports[`Should allow working with vector data Should calculate start index Should calculate for vector data XYZW with typed array 3`] = `12`;
